Signal/src/ViewControllers/OWS2FARegistrationViewController.m

```diff
@@ -0,0 +1,291 @@
+                              selector:@selector(submitButtonWasPressed)];
+    self.submitButton = submitButton;
+    [self.view addSubview:self.submitButton];
+    [self.submitButton autoSetDimension:ALDimensionHeight toSize:kSubmitButtonHeight];
+}
+
+- (CGFloat)hMargin
+{
+    return 20.f;
+}
+
+- (void)createContents
+{
+    const CGFloat kVSpacing = 30.f;
+
+    NSString *instructionsText = NSLocalizedString(
+        @"REGISTER_2FA_INSTRUCTIONS", @"Instructions to enter the 'two-factor auth pin' in the 2FA registration view.");
+    UILabel *instructionsLabel = [self createLabelWithText:instructionsText];
+    [instructionsLabel autoPinTopToSuperviewWithMargin:kVSpacing];
+    [instructionsLabel autoPinWidthToSuperviewWithMargin:self.hMargin];
+
+    UILabel *createForgotLink = [self createForgotLink];
+    [createForgotLink autoPinEdge:ALEdgeTop toEdge:ALEdgeBottom ofView:instructionsLabel withOffset:5];
+    [createForgotLink autoPinWidthToSuperviewWithMargin:self.hMargin];
+
+    [self createPinTextfield];
+    [self.pinTextfield autoPinEdge:ALEdgeTop toEdge:ALEdgeBottom ofView:createForgotLink withOffset:kVSpacing];
+    [self.pinTextfield autoPinWidthToSuperviewWithMargin:self.hMargin];
+
+    UIView *underscoreView = [UIView new];
+    underscoreView.backgroundColor = [UIColor colorWithWhite:0.5 alpha:1.f];
+    [self.view addSubview:underscoreView];
+    [underscoreView autoPinEdge:ALEdgeTop toEdge:ALEdgeBottom ofView:self.pinTextfield withOffset:3];
+    [underscoreView autoPinWidthToSuperviewWithMargin:self.hMargin];
+    [underscoreView autoSetDimension:ALDimensionHeight toSize:1.f];
+
+    [self createSubmitButton];
+    [self.submitButton autoPinEdge:ALEdgeTop toEdge:ALEdgeBottom ofView:underscoreView withOffset:kVSpacing];
+    [self.submitButton autoPinWidthToSuperviewWithMargin:self.hMargin];
+
+    [self updateEnabling];
+}
+
+- (void)updateEnabling
+{
+    [self.submitButton setEnabled:self.hasValidPin];
+}
+
+#pragma mark - UITextFieldDelegate
+
+- (BOOL)textField:(UITextField *)textField
+    shouldChangeCharactersInRange:(NSRange)range
+                replacementString:(NSString *)insertionText
+{
+
+    [ViewControllerUtils ows2FAPINTextField:textField
+              shouldChangeCharactersInRange:range
+                          replacementString:insertionText];
+
+    [self updateEnabling];
+
+    return NO;
+}
+
+#pragma mark - Events
+
+- (void)submitButtonWasPressed
+{
+    OWSAssert(self.hasValidPin);
+
+    [self tryToRegister];
+}
+
+- (BOOL)hasValidPin
+{
+    return self.pinTextfield.text.length >= kMin2FAPinLength;
+}
+
+- (void)tryToRegister
+{
+    OWSAssert(self.hasValidPin);
+    OWSAssert(self.verificationCode.length > 0);
+    NSString *pin = self.pinTextfield.text;
+    OWSAssert(pin.length > 0);
+
+    DDLogInfo(@"%@ %s", self.logTag, __PRETTY_FUNCTION__);
+
+    __weak OWS2FARegistrationViewController *weakSelf = self;
+
+    [ModalActivityIndicatorViewController
```

Comment:
Do we use this progress UI on any of the other steps in registration flow? This feels like an outlier.

Reply:
We don't yet, but I am actually about to add a card suggesting we do that.  IMO we should use the same UI pattern everywhere in our app and `ModalActivityIndicatorViewController` has a bunch of advantages over what we have now in the registration flow.

* Easier for developers to use.
* Is more consistent with how other apps do blocking or modal activity indicators like WA.
* Our registration views use "in-button" activity indicators that can lead to broken layout for buttons with long labels (e.g. issues will only show up in some l10ns).